### ISSUE TITLE
.NET 8: Tracking enable repo-level prebuilt detection

### ISSUE THREAD

**Issue by mmitche** [dotnet/source-build#3017]
## Describe the Problem

Enable prebuilt detection in each repo. Time estimates are provided for the to-do repositories in T-shirt size format (from least work to most work: S, M, L)

### Toolsets

- [x] [source-link](https://github.com/dotnet/sourcelink/pull/933) (opened on 01/02/2023)
- [x] [source-build-reference-packages](https://github.com/dotnet/source-build/issues/2423) (merged 10/16/2021)
- [x] [arcade (M)](https://github.com/dotnet/arcade/pull/13027) (merged 04/07/2023)

### Tier 1

- [x] [command-line-api](https://github.com/dotnet/command-line-api/pull/2012) (merged 01/11/2023)
- [x] [diagnostics](https://github.com/dotnet/diagnostics/pull/3583) (merged 01/05/2023)
- [x] [emsdk](https://github.com/dotnet/emsdk/blob/main/eng/SourceBuildPrebuiltBaseline.xml#L3) (completed!)
- [x] [razor](https://github.com/dotnet/razor/issues/8491) (opened on 3/22/3023)
- [x]  [xliff-tasks](https://github.com/dotnet/source-build/issues/2201) (merged 08/26/2022)
- [x] [roslyn (L)](https://github.com/dotnet/roslyn/pull/67645) (opened 04/05/2023)
- [x] [source-build-externals](https://github.com/dotnet/source-build-externals/pull/22)
- [x] [symreader](https://github.com/dotnet/symreader/pull/274) (opened on 01/04/2023)
- [x] [xdt](https://github.com/dotnet/xdt/issues/469) (merged 10/22/2022)
- [x] [cecil](https://github.com/dotnet/runtime/issues/84985) (merged 4/18/2023)

### Tier 2

- [x] ~linker~ ([planned](https://github.com/dotnet/runtime/issues/75278) to be merged into runtime)
- [x] [runtime](https://github.com/dotnet/runtime/pull/86151) (merged on 05/18/2023)
- [x] [msbuild (M)](https://github.com/dotnet/msbuild/pull/8678) (opened 04/18/2023)

### Tier 3

- [x] [roslyn-analyzers](https://github.com/dotnet/roslyn-analyzers/issues/6606) (M)

### Tier 4

- [x] [aspnetcore (M)](https://github.com/dotnet/aspnetcore/pull/47894) (opened 4/25/2023)
- [x] [deployment-tools](https://github.com/dotnet/deployment-tools/issues/264) (merged 03/17/2023)
- [x] [format](https://github.com/dotnet/format/issues/1809) (opened 3/17/2023)
- [x] [nuget-client (L)](https://github.com/NuGet/Home/issues/12598) (opened on 5/18/2023)
- [x] [templating](https://github.com/dotnet/templating/pull/5582) (merged 01/06/2023)
- [x] [test-templates (S)](https://github.com/dotnet/test-templates/pull/298) (opened 4/25/2023)

### Tier 5

- [x] [fsharp (L)](https://github.com/dotnet/fsharp/issues/15166) (merged 5/16/2023)
- [x] [sdk (L)](https://github.com/dotnet/sdk/pull/31341) (99% done aside from some hangers-on)
- [x] [vstest (M)](https://github.com/microsoft/vstest/issues/4405) merged 5/5/2023)

### Tier 6

- [x] [installer (M)](https://github.com/dotnet/installer/issues/16319) (merged 5/4/2023)

**Comment by mmitche** [comment]
@MichaelSimons Does an issue exist for this in each repo yet (I can't find out). Or a tracking issue (also can't find one). If not I'll make them.

**Comment by MichaelSimons** [comment]
> @MichaelSimons Does an issue exist for this in each repo yet (I can't find out). Or a tracking issue (also can't find one). If not I'll make them.

They do exist but in the source-build repo, not in the individual repos - https://github.com/dotnet/source-build/issues?q=is%3Aissue+is%3Aopen+%5BArPow+Stage+4%5D.  I propose we close these and open repos specific ones.

**Comment by MichaelSimons** [comment]
FYI, source-build-externals, source-build-reference-packages, and xliff-tasks have prebuilt detection already enabled.

**Comment by MichaelSimons** [comment]
[Triage] I will leave this open as a tracking item.  @mmitche, please open repo specific issues and close these -  https://github.com/dotnet/source-build/issues?q=is%3Aissue+is%3Aopen+%5BArPow+Stage+4%5D

**Comment by oleksandr-didyk** [comment]
When enabling pre-build detection on `sourcelink` repository, `MSBuild` version was updated to latest which in turn pulls in `6.0.x` versions of certain runtime libraries. The libraries then are marked as pre-builds, failing the build. `MSBuild` dependencies cannot be bumped to latest since the repo also ships with VS.

A straight-forward solution would be to add reference packages for the libraries in question. However, this solution might not be the best due to several factors:

- `MSBuild` and a few other repositories in the VS eco-system are updating their references to 7.0.x, meaning that the introduced reference packages would need to be updated again after a short period of time
- Product build would still diverge from the repository build, since the latter would use 6.0 / 7.0 versions, while the product one would inject latest versions of the libraries. This has the risk of us discovering avoidable issues late in the product construction process
- We are monitoring use of reference packages for every release to identify obsolete ref packs (mostly due to repositories bumping versions of their dependencies). If a ref pack is repo build only, when we check for its usage in the product build it would be marked as unused. 
- We might run into similar issues down the line, potentially bloating maintenance work around ref packs

Another approach might be to introduce repo PvP before enabling pre-builds. With it implemented, we could use SBRP both in repo builds as well as product build (since product build would consume repo's versions instead of overwriting them with latest as it does now), bringing the two closer. Furthermore, repo build ref packs would also be picked up by the ref pack monitoring, guaranteeing that we would not remove a used ref pack by accident.

While the choice between the two solutions described above is being discussed, we can unblock `sourcelink` pre-build detection work by adding specific exclusions to `SourceBuildPrebuiltBaseline.xml` for the runtime libraries coming from `MSBuild`

**Comment by tkapin** [comment]
Looking at the list of repos, it's not clear how difficult they are to address. Obviously, bigger and more complex repos with more dependencies are going to be more challenging, but it's not apparent from the list itself.

To get clarity about progress expectations, would it be possible to provide a rough estimation (e.g., in t-shirt sizes (S, M, L, XL)) for each of the repos so it's easier to reason about how difficult we think they are to address? /cc @oleksandr-didyk @crummel  @MichaelSimons

**Comment by MichaelSimons** [comment]
@oleksandr-didyk - I see runtime has removed their dependency on Linker but it appears to still be redistributed by the SDK.  Please follow up to find out what the plan is.  This makes me think runtime removed it's dependency but it still needs to be source-built because of the SDK.

**Comment by mmitche** [comment]
> @oleksandr-didyk - I see runtime has removed their dependency on Linker but it appears to still be redistributed by the SDK. Please follow up to find out what the plan is. This makes me think runtime removed it's dependency but it still needs to be source-built because of the SDK.

@MichaelSimons I think its sources may have gotten integrated into runtime, but I could be wrong about that. @lewing may know more.

**Comment by MichaelSimons** [comment]
> > @oleksandr-didyk - I see runtime has removed their dependency on Linker but it appears to still be redistributed by the SDK. Please follow up to find out what the plan is. This makes me think runtime removed it's dependency but it still needs to be source-built because of the SDK.
> 
> @MichaelSimons I think its sources may have gotten integrated into runtime, but I could be wrong about that. @lewing may know more.

Thanks @mmitche.  I ran a source-build locally removing the linker repo.  When I did that I started seeing the following prebuilt from https://github.com/dotnet/sdk/blob/main/src/Layout/redist/redist.csproj#L37.

<Usage Id="Microsoft.NET.ILLink.Analyzers" Version="8.0.100-1.23055.2" File="src/sdk/artifacts/source-build/self/src/artifacts/obj/redist/project.assets.json" IsDirectDependency="true" />

It looks like the issue here is the ILLink components in runtime are built as part of the source-build configuration - https://github.com/dotnet/runtime/blob/main/eng/Subsets.props#L38.  I will open a runtime issue for this.

**Comment by lewing** [comment]
Move is in progress https://github.com/dotnet/runtime/issues/75278 cc @tlakollo

**Comment by MichaelSimons** [comment]
@lewing, @tlakollo - How would you like to track ensuring the ILLink components are built as part of source-build?  Should I log a new issue or add a comment to https://github.com/dotnet/runtime/issues/75278?

**Comment by tlakollo** [comment]
I will create a separate issue and then reference it from https://github.com/dotnet/runtime/issues/75278. Currently, I'm looking into getting a leg to produce the illink packages from runtime and then change the flow from dotnet/linker to dotnet/runtime. When this happens I will need to change the source build info in SDK and installer.

**Comment by oleksandr-didyk** [comment]
@tkapin as you have requested, added time estimates in T-shirt sizes with the help from Michael. Additionally, certain repositories might take priority due to the amount of pre-built issues they are currently causing

From the top of my head, these are:
  - runtime
  - sdk
  - format
 
I most probably missed 1-2, @MichaelSimons can maybe expand the list more

**Comment by tkapin** [comment]
Thank you @oleksandr-didyk, this helps to reason about difficulty of each repos.

**Comment by oleksandr-didyk** [comment]
Update on progress: 
  - `sourcelink` and `deployment-tools` repos currently waiting on a UX PR to Arcade that simplifies usage of rid-specific intermediate packages -> https://github.com/dotnet/arcade/pull/12235
  - working on pre-built detection in `runtime` while the PR is being reviewed

**Comment by oleksandr-didyk** [comment]
Update on progress::
  - `runtime` is waiting on several packages to be added to SBRP; ran into issues with SBRPing for `net6.0`, hopefully will be fixed with https://github.com/dotnet/source-build-reference-packages/pull/547
  - [`format`](https://github.com/dotnet/format/issues/1802) and [`sourcelink`](https://github.com/dotnet/sourcelink/pull/936) are waiting on upgrades to `$(NetLatest)` (`net8.0`) for bumping several packages to latest / using Arcade's improvements for source-build: https://github.com/dotnet/arcade/issues/12222 and https://github.com/dotnet/source-build/issues/3296

Next repos for pre-build detection will probably be:
  1. roslyn
  2. msbuild
  3. roslyn-analyzers
  4. nuget-client
  5. aspnetcore

**Comment by mthalman** [comment]
I noticed [emsdk](https://github.com/dotnet/emsdk) wasn't being tracked in the list so I've added that.

**Comment by mmitche** [comment]
BAM! It's already done.

**Comment by MichaelSimons** [comment]
> BAM! It's already done.

Thanks to @directhex who enabled prebuilt detection when he onboarded the repo to source-build.

**Comment by mthalman** [comment]
I've added [cecil](https://github.com/dotnet/cecil) to the list since that's been moved to its own repo.
